descqa/StellarMassFunction.py

```diff
@@ -0,0 +1,353 @@
+
+        results={}
+        if all(x in validation_data.keys() for x in ('logM', 'log_phi', 'dlog_phi+', 'dlog_phi-')):
+            M = np.power(10, validation_data['logM'])
+            phi = np.power(10, validation_data['log_phi'])
+            dphi_hi = np.power(10, validation_data['log_phi'] + validation_data['dlog_phi+']) - phi
+            dphi_lo = -np.power(10, validation_data['log_phi'] + validation_data['dlog_phi-']) + phi
+            ax.errorbar(M, phi, yerr=[dphi_lo,dphi_hi], color=self.validation_color, marker=self.validation_marker, label=validation_label, ms=self.msize)
+            results['data']=phi
+            results['Mdata']=M
+            results['data+'] = dphi_hi
+            results['data-'] = dphi_lo
+        else:
+            raise ValueError("Missing expected validation-data quantitites")
+
+        return results
+
+
+    def decorate_subplot(self, ax, nplot, label=None):
+        ax.set_xscale('log')
+        ax.set_yscale('log')
+        if label:
+            ax.text(0.95, 0.95, label, horizontalalignment='right', verticalalignment='top', transform=ax.transAxes) 
+
+        #add axes and legend
+        if nplot+1 <= self.nplots-self.ncolumns:  #x scales for last ncol plots only
+            #print "noticks",nplot
+            for axlabel in ax.get_xticklabels():
+                axlabel.set_visible(False)
+                #prevent overlapping yaxis labels
+                ax.yaxis.get_major_ticks()[0].label1.set_visible(False)
+        else:
+            ax.set_xlabel(self.xaxis)
+            for axlabel in ax.get_xticklabels():
+                axlabel.set_visible(True)
+        ax.legend(loc='best', fancybox=True, framealpha=0.5, numpoints=1)
+
+
+    @staticmethod
+    def post_process_plot(fig):
+        fig.subplots_adjust(hspace=0)
+
+
+    @staticmethod
+    def save_quantities(keyname, results, filename, comment=''):
+        if keyname in results:
+            if keyname+'-' in results and keyname+'+' in results:
+                fields = ('M'+keyname, keyname, keyname+'-', keyname+'+')
+                header = ', '.join(('Data columns are: <M>', keyname, keyname+'-', keyname+'+', ' '))
+            elif keyname+'+-' in results:
+                fields = ('M'+keyname, keyname, keyname+'+-')
+                header = ', '.join(('Data columns are: <M>', keyname, keyname+'+-',' '))
+            else:
+                fields = ('M'+keyname, keyname)
+                header = ', '.join(('Data columns are: <M>', keyname, ' '))
+            np.savetxt(filename, np.vstack((results[k] for k in fields)).T, fmt='%12.4e', header=header+comment)
+
+
+    def conclude_test(self, output_dir):
+        self.post_process_plot(self.summary_fig)
+        self.summary_fig.savefig(os.path.join(output_dir, 'summary.png'))
+        plt.close(self.summary_fig)
+
+
+    @staticmethod
+    def get_sky_volume(catalog_instance, zhi, zlo=0., distance='Mpc'):
```

Comment:
`distance` is not used in this function

Reply:
OK I changed distance to units and added a check for units=='Mpc'. If this is going to be a function in a utilities module, we could have something more bulletproof than this. Lots of test developers will need a "get_sky_volume" function. It would save everyone doing it for themselves and will help with minimizing errors if we offer it, I think.